examples/pytorch/graphsage/lightning/node_classification.py

```diff
@@ -0,0 +1,145 @@
+            g.ndata['h'] = y
+        return y
+
+    def training_step(self, batch, batch_idx):
+        input_nodes, output_nodes, blocks = batch
+        x = blocks[0].srcdata['feat']
+        y = blocks[-1].dstdata['label']
+        y_hat = self(blocks, x)
+        loss = F.cross_entropy(y_hat, y)
+        self.train_acc(torch.softmax(y_hat, 1), y)
+        self.log('train_acc', self.train_acc, prog_bar=True, on_step=True, on_epoch=False)
+        return loss
+
+    def validation_step(self, batch, batch_idx):
+        input_nodes, output_nodes, blocks = batch
+        x = blocks[0].srcdata['feat']
+        y = blocks[-1].dstdata['label']
+        y_hat = self(blocks, x)
+        self.val_acc(torch.softmax(y_hat, 1), y)
+        self.log('val_acc', self.val_acc, prog_bar=True, on_step=True, on_epoch=True, sync_dist=True)
+
+    def configure_optimizers(self):
+        optimizer = torch.optim.Adam(self.parameters(), lr=0.001, weight_decay=5e-4)
+        return optimizer
+
+class DataModule(LightningDataModule):
+    def __init__(self, graph, train_idx, val_idx, fanouts, batch_size, n_classes):
+        super().__init__()
+
+        sampler = dgl.dataloading.NeighborSampler(
+            fanouts, prefetch_node_feats=['feat'], prefetch_labels=['label'])
+
+        self.g = graph
+        self.train_idx, self.val_idx = train_idx, val_idx
+        self.sampler = sampler
+        self.batch_size = batch_size
+        self.in_feats = graph.ndata['feat'].shape[1]
+        self.n_classes = n_classes
+
+    def train_dataloader(self):
+        return dgl.dataloading.DataLoader(
+            self.g, self.train_idx.to('cuda'), self.sampler,
+            device='cuda', batch_size=self.batch_size, shuffle=True, drop_last=False,
+            # For CPU sampling, set num_workers to nonzero and use_uva=False
+            # Set use_ddp to False for single GPU.
+            num_workers=0, use_uva=True, use_ddp=True)
+
+    def val_dataloader(self):
+        return dgl.dataloading.DataLoader(
+            self.g, self.val_idx.to('cuda'), self.sampler,
+            device='cuda', batch_size=self.batch_size, shuffle=True, drop_last=False,
+            num_workers=0, use_uva=True)
+
+if __name__ == '__main__':
+    dataset = DglNodePropPredDataset('ogbn-products')
+    graph, labels = dataset[0]
+    graph.ndata['label'] = labels.squeeze()
+    split_idx = dataset.get_idx_split()
+    train_idx, val_idx, test_idx = split_idx['train'], split_idx['valid'], split_idx['test']
+    datamodule = DataModule(graph, train_idx, val_idx, [15, 10, 5], 1024, dataset.num_classes)
+    model = SAGE(datamodule.in_feats, 256, datamodule.n_classes)
+
+    # Train
+    checkpoint_callback = ModelCheckpoint(monitor='val_acc', save_top_k=1)
+    # Use this for single GPU
+    #trainer = Trainer(gpus=[0], max_epochs=10, callbacks=[checkpoint_callback])
+    trainer = Trainer(gpus=[0, 1, 2, 3], max_epochs=10, callbacks=[checkpoint_callback], strategy='ddp_spawn')
```

Comment:
@justusschock I tried `ddp` and while it runs indeed faster, the code after `trainer.fit` has been run 4 times for multi-GPU.  Is this expected for PyTorch Lightning?

Reply:
We rerun the whole script, so yes to some extend this is expected. Approaches like `trainer.test` or `trainer.predict` also take the existing processes to setup distributed inference. `ddp_spawn` only runs the current function in case that is crucial for you.

Reply:
Noted.  Thanks!